src/dmap_import/api_copy_job.py

```diff
@@ -0,0 +1,190 @@
+import os
+import gzip
+import datetime
+from typing import Any
+
+import sqlalchemy as sa
+from sqlalchemy.dialects import postgresql
+from sqlalchemy.engine import CursorResult
+
+from dmap_import.schemas.api_metadata import ApiMetadata
+from dmap_import.util_api import (
+    download_from_url,
+    get_api_results,
+    ApiResult,
+)
+from dmap_import.util_rds import (
+    copy_local_to_db,
+    DatabaseManager,
+)
+from dmap_import.util_logging import ProcessLogger
+
+LAST_UPDATED_FORMAT = "%Y-%m-%dT%H:%M:%S.%f"
+
+
+def schema_compare(local_path: str, destination_table: Any) -> None:
+    """
+    Compare columns of local file, downloaded from API, to expected columns
+    from SQLAlchemy schema.
+
+    throw if columns in local file don't exist in expected schema
+
+    if columns in expected schema, but not in local file, just log
+
+    :param local_path: local file path of API dataset file
+    :param destination_table: SQLAlchemy table object to compare
+    """
+    schema_compare_log = ProcessLogger(
+        "schema_compare",
+        local_path=local_path,
+        destination_table=str(destination_table.__table__),
+    )
+    schema_compare_log.log_start()
+
+    table = destination_table.__table__
+    ignore_columns = (
+        "pk_id",
+        "dataset_id",
+    )
+    # table.columns object produces column names with table name pre-pended
+    # must remove table pre-pend for comparison
+    destination_columns = [
+        str(col).replace(f"{table}.", "") for col in table.columns
+    ]
+    # drop ignore_columns from comparison list, as they won't exist in
+    # downloaded file
+    destination_columns = [
+        col for col in destination_columns if col not in ignore_columns
+    ]
+
+    with gzip.open(local_path, "rt") as gzip_file:
+        local_columns = gzip_file.readline().strip().lower().split(",")
+
+    # columns in defined schema, but not in local file
+    not_in_local = set(destination_columns).difference(set(local_columns))
+
+    # column in local file, but not in defined schema
+    not_in_dest = set(local_columns).difference(set(destination_columns))
+
+    schema_compare_log.add_metadata(
+        local_col_count=len(local_columns),
+        dest_col_count=len(destination_columns),
+        not_in_local_count=len(not_in_local),
+        not_in_dest_count=len(not_in_dest),
+        not_in_local_columns=" | ".join(not_in_local),
+        not_in_dest_columns=" | ".join(not_in_dest),
+    )
+
+    if not_in_dest:
+        exception = IndexError(
+            (
+                f"Columns: '{' | '.join(not_in_dest)}' "
+                f"found in {local_path} but not in {table}"
+            )
+        )
+        schema_compare_log.log_failure(exception=exception)
+        raise exception
```

Comment:
we'll catch this later when running `COPY` but we could also check for missing required columns in dest that are missing from source here.